feat: convert functions to v4

pudkrong opened the pull request:
In response to the absence of the `function.json` file in programming model 4, we will now use the `handler.ts` file as the initial point for the Azure Function. This file will take on the responsibility of registering the function in Azure, following the modifications outlined in the `package.json` file, with particular attention to the `main` property. It's worth noting that during the packaging and deployment process, the file extension will be changed to `handler.mjs`

Ref: https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-upgrade-v4?tabs=v4

pudkrong committed:
feat: migrate signalR functions

coderbyheart commented on agnssDeviceRequestsHandler/agnssDeviceRequestsHandler.ts:
```diff
@@ -29,21 +29,14 @@ const config = () =>
+) => {
 	log(context)({ context, requests })
 
+	if (!Array.isArray(requests)) return
+
```
**question:** why protect against this case?

pudkrong replied:
According to https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/96cda12e50219c2d1761b665ecc5cff67009eee0/agnssDeviceRequestsHandler/agnssDeviceRequestsHandler.ts#L58, we need to assert the type.

coderbyheart replied:
This can be solved by typing the handler arguments correctly.

pudkrong replied:
Argument has type as unknown, when I tried to type it. It errors as

```
 Type 'unknown' is not assignable to type {Your type}
```

So, I think the best way is to validate the input using TypeBox. But I think it would be unrelated to this PR, unless you think it should.

coderbyheart replied:
Then the type is wrong, because we know what the input is.

coderbyheart commented on agnssQueuedDeviceRequestsHandler/handler.ts:
```diff
@@ -0,0 +1,8 @@
+
+app.storageQueue('agnssQueuedDeviceRequestsHandler', {
+	queueName: '%AGNSS_REQUESTS_QUEUE_NAME%',
+	connection: '',
```
**question:** Why is this empty?

pudkrong replied:
`connection` is required by `storageQueue` type. Leaving it empty, Azure will use the default storage connection as per https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-queue-trigger?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cextensionv5&pivots=programming-language-typescript#connection-string

coderbyheart commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "description": "A reference implementation of a serverless backend for an IoT product developed using in TypeScript.",
   "type": "module",
-  "main": "./dist/export.js",
+  "main": "./dist/**/handler.mjs",
```
**problem**: Main field can't have wildcards.

pudkrong replied:
This is recommended by Microsoft, https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-upgrade-v4?tabs=v4#set-your-app-entry-point

coderbyheart replied:
That's weird, and Azure specific behaviour. So instead of hacking the package.json, I'd like to see this added to the `host.json` instead.

pudkrong replied:
I couldn't find a way to register functions via `host.json`. So, you have agreed to use wild card on `main` field for now.

pudkrong committed:
feat: use proper type

pudkrong committed:
feat: convert functions to v4

Migrate to programming model v4.
Ref: https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-upgrade-v4?tabs=v4

coderbyheart commented on agnssDeviceRequestsHandler/agnssDeviceRequestsHandler.ts:
```diff
@@ -63,17 +63,21 @@ const agnssDeviceRequestsHandler: AzureFunction = async (
-		}))
+		.map((request, i) => {
+			const systemPropertiesArray = Array.isArray(
+				context.triggerMetadata?.systemPropertiesArray,
```
Why is the check on the array property needed whether it is an array?

pudkrong replied:
The `context` is `InvocationContext` type where `triggerMetadata` has `Record<string, unknown>` type. Therefore, `systemPropertiesArray` is unknown type. 

I think assertion here is needed.

coderbyheart replied:
But we know this property is an array, it IS in the name: `systemPropertiesArray`. So the type needs to be corrected.

coderbyheart commented on agnssDeviceRequestsHandler/agnssDeviceRequestsHandler.ts:
```diff
@@ -66,13 +76,10 @@ const agnssDeviceRequestsHandler: AzureFunction = async (
 			request,
 			deviceId:
-				context.bindingData.systemPropertiesArray[i][
+				context.triggerMetadata.systemPropertiesArray?.[i][
```
```suggestion
				context.triggerMetadata.systemPropertiesArray[i][
```

Always present.

pudkrong committed:
fix: using the right type

EventHub context has `triggerMetadata` property where contains event meta data values.
In the triggerMetadata property, there are 2 properties we are interested.

- `propertiesArray`
  - User properties of the event data that the user explicitly added during send operations
  - It presents as non-empty array
  - If user does not explicitly add, it will be empty object

    ```
    "propertiesArray: [
      {},
      {}
    ]
    ```

- `systemPropertiesArray`
  - It is always present
